### PR TITLE
[#250 clone] Run `ktlint` in process isolation + allow overriding `ktlint` version

### LINKED ISSUE #34: Is that possible to include custom rule from current project?
Readme states two ways how to add custom rules
```gradle
buildscript {
    dependencies {
        classpath files('libs/my-custom-ktlint-rules.jar')
        classpath 'org.other.ktlint:custom-rules:1.0'
    }
}
```

I'd like to have custom rules as separate module in my project. Is that possible to include them directly? Ideally, I'd like that these rules are applied against themselves as well.

### LINKED ISSUE #280: If we can provide a property in `KotlinterExtension` to override ktlint version?
The current way for users is a little inconvenient.

### PR DESCRIPTION
**Pull request by mateuszkwiecinski**
Clone of #250 with my suggested improvements/changes

1. new configurations will be automatically configured, even for custom tasks
2. It's possible to override `ktlint` version
3. @hrach's custom rules were migrated to new ktlint 0.47.1 apis
4. I slightly changed the way errors are reported back
5. Plugin itself won't expose `ktlint` dependency anymore to its consumer (it's defined as `compileOnly` now)
6. default runtime ktlint version is kept in sync with the one compilation uses
7. [new] It doesn't only uses `ktlint` APIs, but also resolves rules from the `ktlint` version provided via plugin extension

Fixes #280
Fixes #34

### CODE REVIEW

**Comment by Karn** [comment]
Hey @mateuszkwiecinski! Looking forward to this PR! I'm wondering if you have considered just manually loading the custom rules using something like:

```kotlint
fun ktlintRulesetsFromClasspath(classpath: ConfigurableFileCollection): List<RuleSetProviderV2> {
    // Load the files from the classpath into a new ClassLoader
    @Suppress("DEPRECATION")
    val fileUris = classpath.map { it.toURL() }.toTypedArray()
    val classLoader = URLClassLoader(fileUris, Thread.currentThread().contextClassLoader)
    return ServiceLoader.load(RuleSetProviderV2::class.java, classLoader).toList()
}
```

Where `classpath` here is the `ktlintClasspath`. That way we can bypass the processIsolation issues, at least temporarily.

**Comment by mateuszkwiecinski** [comment]
@Karn Thanks for sharing the idea💡 
I haven't considered any workarounds yet, mainly because the process isolation works just fine. I did some manual testing, and it all worked like a charm. The blocker I stumbled upon turned out not being related to the process isolation, but rather the quirky behavior of Gradle TestKit testing framework 🤷 
I disabled problematic tests on Windows, and the PR soon reach reviewable state :)

File: src/test/kotlin/org/jmailen/gradle/kotlinter/functional/KotlinProjectTest.kt [hunk 1]
```diff
@@ -186,6 +186,30 @@ internal class KotlinProjectTest : WithGradleTest.Kotlin() {
...
+
+        build("lintKotlin", "--info").apply {
+            assertEquals(SUCCESS, task(":lintKotlin")?.outcome)
+            assertTrue(output.contains("Resolved 45 RuleSetProviders"))
+        }
+
+        build("formatKotlin", "--info").apply {
+            assertEquals(SUCCESS, task(":formatKotlin")?.outcome)
+            assertTrue(output.contains("Resolved 45 RuleSetProviders"))
+        }
```
Comment (mateuszkwiecinski) [inline src/test/kotlin/org/jmailen/gradle/kotlinter/functional/KotlinProjectTest.kt:210]: This is not ideal tests, and will likely fail when upgrading ktlint version, but I was out of ideas and decided to defer writing better assertion to somewhen _later_ 😬 
What I want to test here is to somehow replace the old test relying on test classpath, with something that covers the dynamic loading part.

File: test-project-android/app/build.gradle.kts [hunk 2]
```diff
@@ -6,6 +6,7 @@ plugins {
 
 android {
     compileSdkVersion(33)
+    namespace = "com.kotlinter.example"
```
Comment (mateuszkwiecinski) [inline test-project-android/app/build.gradle.kts:9]: Required by AGP 8.0, setting package name via AndroidManifest results in an error now

File: README.md [hunk 3]
```diff
@@ -157,6 +157,7 @@ kotlinter {
     reporters = arrayOf("checkstyle", "plain")
     experimentalRules = false
     disabledRules = emptyArray()
+    ktlintVersion = "x.y.z"
```
[the remark this answers is not in the document]
Comment (mateuszkwiecinski) [reply README.md:160]: As an alternative, I can propose having some sort of hardcoded value that will be mapped into "our" ktlint version, i.e.
```suggestion
    ktlintVersion = "default" // or `latest` or `current`? but that might be too confusing
```

File: README.md [hunk 4]
```diff
@@ -188,6 +190,9 @@ disabledRules = ["no-wildcard-imports"]
 ```
 You must prefix rule ids not part of the standard rule set with `<rule-set-id>:<rule-id>`. For example `experimental:annotation`.
 
+There is a basic support for overriding `ktlintVersion`, but the plugin doesn't guarantee backwards compatibility with all `ktlint` versions.
+Errors like `java.lang.NoSuchMethodError:` or `com/pinterest/ktlint/core/KtLint$Params` can be thrown if provided `ktlint` version isn't compatible with the latest ktlint apis.
```
[the remark this answers is not in the document]
Comment (jeremymailen) [reply README.md:194]: Backwards compatibility, no, but I think the main use case for selecting a version of ktlint is if someone wants to upgrade to a newer version of ktlint when kotlinter hasn't yet been upgraded -- assuming there aren't any changes to compatibility, which often there aren't.

File: build.gradle.kts [hunk 5]
```diff
@@ -59,16 +59,15 @@ dependencies {
         "ktlint-reporter-json",
         "ktlint-reporter-html",
         "ktlint-reporter-plain",
-        "ktlint-reporter-sarif",
-        "ktlint-ruleset-experimental",
-        "ktlint-ruleset-standard"
```
[the remark this answers is not in the document]
Comment (mateuszkwiecinski) [reply build.gradle.kts:64]: >Ideally we'd only need ktlint-core but due to the fact reporters leaked to our implementation I decided to leave them as it is, opening us for bugs in classpath missmatch, but AFAIU this should only happen if reporters api changes.

I actually managed to clean the rest of the dependencies already, in https://github.com/jeremymailen/kotlinter-gradle/pull/294

**Comment by hrach** [comment]
Friendly reminder - it would be great to have the functionality for custom rules. Thank you.

**Comment by jeremymailen** [comment]
Thanks, this example was very helpful. The feature has been incorporated in 5.0.0. Cheers

**Pull request closed without merging**